src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala

```diff
@@ -95,16 +98,39 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     }
     def checkCq(cq: ConjunctiveQuery) {
       checkAtom(cq.head)
-      cq.bodies foreach { x => x foreach checkAtom }
+      checkBodyAtoms(checkAtom)(cq.bodies.flatten)
     }
     stmt match {
       case s: ExtractionRule => checkCq(s.q)
       case s: InferenceRule  => checkCq(s.q)
-      case _ => 
+      case _ =>
+    }
+  }
+
+  // check if outer join body contains one atom
+  def checkOptionalModifier(stmt: Statement) {
+    def checkOuterJoinBodies(bodies: List[Body]) {
+      bodies.foreach {
+        case b: QuantifiedBody => {
+          b.modifier match {
+            case OuterModifier() =>
+              if ((b.bodies collect { case x: Atom => 1 }).size != 1) {
+                error(stmt, s"One and only one atom should be supplied in OPTIONAL modifier")
+              }
+            case _ =>
+          }
+        }
+        case _ =>
+      }
+    }
+    stmt match {
+      case s: ExtractionRule => checkOuterJoinBodies(s.q.bodies.flatten)
+      case s: InferenceRule => checkOuterJoinBodies(s.q.bodies.flatten)
```

Comment:
I suggest avoiding `flatten` to simply call this type of scanning functions.  Unless you want to retain the flattened list, I think it's better to use iterators such as `foreach` and avoid creating temporary lists.